README.rst
kotti_google_analytics
************************

This extension allows Google Analytics to Kotti, in order to track users and sessions on your site.
In addition, it adds Google Analytics Visualizations for reporting by using
`Google Analytics Embed API`_.

Example of Google Analytics Visualization:
--------------------------------------------

.. image:: https://static.raymondcamden.com/images/wp-content/uploads/2015/07/shot11.png

This is an extension to Kotti that allows to add foo to your site.

|pypi|_
|downloads_month|_
|license|_
|build_status_stable|_


.. _Google Analytics Embed API: https://ga-dev-tools.appspot.com/

.. |pypi| image:: https://img.shields.io/pypi/v/kotti_google_analytics.svg?style=flat-square
.. _pypi: https://pypi.python.org/pypi/kotti_google_analytics/

.. |downloads_month| image:: https://img.shields.io/pypi/dm/kotti_google_analytics.svg?style=flat-square
.. _downloads_month: https://pypi.python.org/pypi/kotti_google_analytics/

.. |license| image:: https://img.shields.io/pypi/l/kotti_google_analytics.svg?style=flat-square
.. _license: http://www.repoze.org/LICENSE.txt

.. |build_status_stable| image:: https://img.shields.io/travis/b4oshany/kotti_google_analytics/production.svg?style=flat-square
.. _build_status_stable: http://travis-ci.org/b4oshany/kotti_google_analytics

`Find out more about Kotti`_

Development happens at https://github.com/b4oshany/kotti_google_analytics

.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti

Setup
=====


To enable the extension in your Kotti site, activate the configurator::

    kotti.configurators =
	    kotti_controlpanel.kotti_configure
        kotti_google_analytics.kotti_configure
        
    kotti_google_analytics.tracking_id = track_id_here


Database upgrade
================

If you are upgrading from a previous version you might have to migrate your
database.  The migration is performed with `alembic`_ and Kotti's console script
``kotti-migrate``. To migrate, run
``kotti-migrate upgrade --scripts=kotti_google_analytics:alembic``.

For integration of alembic in your environment please refer to the
`alembic documentation`_. If you have problems with the upgrade,
please create a new issue in the `tracker`_.

Development
===========

|build_status_master|_

.. |build_status_master| image:: https://img.shields.io/travis/b4oshany/kotti_google_analytics/master.svg?style=flat-square
.. _build_status_master: http://travis-ci.org/b4oshany/kotti_google_analytics

Contributions to kotti_google_analytics are highly welcome.
Just clone its `Github repository`_ and submit your contributions as pull requests.

.. _alembic: http://pypi.python.org/pypi/alembic
.. _alembic documentation: http://alembic.readthedocs.org/en/latest/index.html
.. _tracker: https://github.com/b4oshany/kotti_google_analytics/issues
.. _Github repository: https://github.com/b4oshany/kotti_google_analytics


CHANGES.rst
History
=======

1.2.0 - 1.2.1
----------------
Date: 2017-06-09

- Added support for Google Analytics USER ID feature
- Pinned `kotti_controlpanel`_ to 1.0.9 or greater

1.1.20
-------

- Fixed instruction links

1.1.19
--------

- Pinned `kotti_controlpanel`_ to 1.0.3 or greater

1.1.18
-------

- Added more charts.

1.1.17
-------

- Pinned `kotti_controlpanel`_ to 1.0.2 or greater

1.1.15 - 1.1.16
-----------------

- Pinned `kotti_controlpanel`_ to 1.0.1 or greater
- Adjusted this package for the new version of `kotti_controlpanel`_.


1.1.11 - 1.1.14
----------------

- FIxed error caused by click

1.1.1 - 1.1.10
-----------------

- Fixed rst files.

1.1.0
-------

- Added Google visualization for reports.
- Added `kotti_controlpanel`_ kotti package, in order to save Google Analytics settings.
- Added `Google Analytics Embed API`_ for visualizations.
- Added `googleanalytics`_ python package to generate access token for Google Analytics.

.. _Google Analytics Embed API: https://ga-dev-tools.appspot.com/
.. _kotti_controlpanel: https://pypi.python.org/pypi/kotti_controlpanel
.. _googleanalytics: https://pypi.python.org/pypi/googleanalytics


1.0.3
------

- Set the tracking code to be placed in the slot below the content area.


1.0.2
-------

- Added Tracking code


1.0.0 - unreleased
--------------------

- Create package with ``pcreate -s kotti kotti_google_analytics``.


i18n.sh
#!/bin/bash
# Usage:
#     Initial catalog creation (lang is the language identifier):
#         ./i18n.sh lang
#     Updating translation and compile catalog:
#         ./i18n.sh

# configuration
DOMAIN="kotti_google_analytics"
SEARCH_PATH="kotti_google_analytics"
LOCALES_PATH="kotti_google_analytics/locale"
# end configuration

# create locales folder if not exists
if [ ! -d "$LOCALES_PATH" ]; then
    echo "Locales directory not exists, create"
    mkdir -p "$LOCALES_PATH"
fi

# create pot if not exists
if [ ! -f "$LOCALES_PATH"/$DOMAIN.pot ]; then
    echo "Create pot file"
    touch "$LOCALES_PATH"/$DOMAIN.pot
fi

# no arguments, extract and update
if [ $# -eq 0 ]; then
    echo "Extract messages"
    pot-create "$SEARCH_PATH" -d $DOMAIN -o "$LOCALES_PATH"/$DOMAIN.pot

    echo "Update translations"
    for po in "$LOCALES_PATH"/*/LC_MESSAGES/$DOMAIN.po; do
        msgmerge --no-wrap -o "$po" "$po" "$LOCALES_PATH"/$DOMAIN.pot
    done

    echo "Compile message catalogs"
    for po in "$LOCALES_PATH"/*/LC_MESSAGES/*.po; do
        echo "Compiling file $po..."
        msgfmt --statistics -o "${po%.*}.mo" "$po"
    done

# first argument represents language identifier, create catalog
else
    cd "$LOCALES_PATH"
    mkdir -p $1/LC_MESSAGES
    msginit -i $DOMAIN.pot -o $1/LC_MESSAGES/$DOMAIN.po -l $1
fi


kotti_google_analytics/populate.py
import colander

from kotti.util import Link
from kotti_controlpanel.util import add_settings
from kotti_controlpanel.util import get_setting
from kotti_google_analytics import _, controlpanel_id, AnayticsDefault


class AnalyticsSchema(colander.MappingSchema):
    client_id = colander.SchemaNode(
        colander.String(),
        name="client_id",
        title=_(u'Client ID'),
    )
    client_secret = colander.SchemaNode(
        colander.String(),
        name="client_secret",
        title=_(u'Client Secret'),
    )
    identity = colander.SchemaNode(
        colander.String(),
        name="identity",
        title=_(u'Identity'),
    )
    access_token = colander.SchemaNode(
        colander.String(),
        name="access_token",
        title=_(u'Access Token'),
    )
    refresh_token = colander.SchemaNode(
        colander.String(),
        name="refresh_token",
        title=_(u'Refresh Token'),
    )


GAControlPanel = {
    'name': controlpanel_id,
    'title': _(u'Google Analytics Settings'),
    'description': _(u"Settings for google_analytics"),
    'success_message': _(u"Successfully saved google_analytics settings."),
    'schema_factory': AnalyticsSchema,
}


def populate():

    links = [
        Link('analytics-report', title=_(u"Google Analytics Report")),
        Link('analytics-setup', title=_(u"Setup Google Analytics"))
    ]

    add_settings(GAControlPanel, links=links)

    AnayticsDefault.client_id = get_setting("client_id")
    AnayticsDefault.client_secret = get_setting("client_secret")
    AnayticsDefault.access_token = get_setting("access_token")
    AnayticsDefault.refresh_token = get_setting("refresh_token")
    AnayticsDefault.identity = get_setting("identity")


kotti_google_analytics/views/view.py
# -*- coding: utf-8 -*-

"""
Created on 2016-06-18
:author: Oshane Bailey (b4.oshany@gmail.com)
"""
import os
import json

from pyramid.view import view_config
from pyramid.view import view_defaults
from pyramid import httpexceptions as httpexc

import googleanalytics as ga
from kotti_controlpanel.util import set_setting, get_setting
from kotti_google_analytics import _, AnayticsDefault
from kotti_google_analytics.fanstatic import css_and_js
from kotti_google_analytics.views import BaseView


@view_config(
    name='analytics-code',
    renderer='kotti_google_analytics:templates/tracking_code.pt')
class CartView(BaseView):

    def __call__(self):
        return {
            "tracking_id": AnayticsDefault.tracking_id
        }


class Analytics(BaseView):

    @property
    def client_id(self):
        return os.environ.get('GOOGLE_ANALYTICS_CLIENT_ID')

    @property
    def client_secret(self):
        return os.environ.get('GOOGLE_ANALYTICS_CLIENT_SECRET')

    @property
    def credentials(self):
        if not hasattr(self, "__credentials"):
            self.__credentials = self.flow.step2_exchange(AnayticsDefault.refresh_token)
        return self.__credentials

    @property
    def flow(self):
        return ga.auth.Flow(
            AnayticsDefault.client_id,
            AnayticsDefault.client_secret,
            redirect_uri=self.request.resource_url(
                self.context, "google-analytics-callback"))

    @property
    def profile(self):
        if not hasattr(self, "__profile"):
            data = {
                "access_token": AnayticsDefault.access_token,
                "client_email": None,
                "refresh_token": AnayticsDefault.refresh_token,
                "client_id": AnayticsDefault.client_id,
                "client_secret": AnayticsDefault.client_secret,
                "identity": AnayticsDefault.identity}
            self.__profile = ga.authenticate(
                account='dpk.alteroo.com',
                webproperty="UA-82860015-1",
                **data
            )
        return self.__profile

    @view_config(name='analytics-report', root_only=True,
                 permission="admin",
                 renderer="kotti_google_analytics:templates/analytics.pt")
    def view(self):
        return {}

    @view_config(name='analytics-setup', root_only=True, permission="admin")
    def setup_analytics(self):
        authorize_url = self.flow.step1_get_authorize_url()
        return httpexc.HTTPFound(location=authorize_url)

    @view_config(name="google-analytics-callback", root_only=True,
                 permission="admin", renderer="json")
    def callback(self):
        credentials = self.flow.step2_exchange(self.request.params['code'])
        jcred = credentials.serialize()
        set_setting("client_id", jcred.get("client_id"))
        set_setting("client_secret", jcred.get("client_secret"))
        set_setting("access_token", jcred.get("access_token"))
        set_setting("refresh_token", jcred.get("refresh_token"))
        set_setting("identity", jcred.get("identity"))
        return jcred


kotti_google_analytics/static/embed.js
// == NOTE ==
// This code uses ES6 promises. If you want to use this code in a browser
// that doesn't supporting promises natively, you'll have to include a polyfill.

setTimeout(function(){
    ga_active_users();
    ga_date_range();
    ga_view_selector();
    gapi_init();
}, 2500);


function getProfiles(cb) {
  //do we have a cached version?
  if(sessionStorage["gaProfiles"]) {
    console.log("profiles fetched from cache");
    cb(JSON.parse(sessionStorage["gaProfiles"]));
    return;
  }

  gapi.client.analytics.management.accounts.list().then(function(res) { 
    var accountId = res.result.items[0].id;
    var profiles = [];
    gapi.client.analytics.management.webproperties.list({'accountId': accountId}).then(function(res) {
  
        res.result.items.forEach(function(item) {
            if(item.defaultProfileId) profiles.push({id:"ga:"+item.defaultProfileId,name:item.name});
        });
      sessionStorage["gaProfiles"] = JSON.stringify(profiles);    
      cb(profiles);      
    });
  });
};


function gapi_init(){
    

    gapi.analytics.ready(function() {
    
      /**
       * Authorize the user immediately if the user has already granted access.
       * If no access has been created, render an authorize button inside the
       * element with the ID "embed-api-auth-container".
       */
    //   gapi.analytics.auth.authorize({
    //     container: 'embed-api-auth-container',
    //     clientid: kga_clientid
    //   });
      
      gapi.analytics.auth.authorize({
        'serverAuth': {
          'access_token': kga_access_token
        }
      });


      gapi.analytics.auth.on('success', function(response) {
        //hide the auth-button
        getProfiles(function(profs) {
            console.log(profs);
          window.profiles = profs;
        //   processProfiles();      
        });
      
      });
    
    
      /**
       * Create a new ActiveUsers instance to be rendered inside of an
       * element with the id "active-users-container" and poll for changes every
       * five seconds.
       */
      var activeUsers = new gapi.analytics.ext.ActiveUsers({
        container: 'active-users-container',
        pollingInterval: 5
      });
    
    
      /**
       * Add CSS animation to visually show the when users come and go.
       */
      activeUsers.once('success', function() {
        var element = this.container.firstChild;
        var timeout;
    
        this.on('change', function(data) {
          var element = this.container.firstChild;
          var animationClass = data.delta > 0 ? 'is-increasing' : 'is-decreasing';
          element.className += (' ' + animationClass);
    
          clearTimeout(timeout);
          timeout = setTimeout(function() {
            element.className =
                element.className.replace(/ is-(increasing|decreasing)/g, '');
          }, 3000);
        });
      });
    
    
      /**
       * Create a new ViewSelector2 instance to be rendered inside of an
       * element with the id "view-selector-container".
       */
      var viewSelector = new gapi.analytics.ext.ViewSelector2({
        container: 'view-selector-container',
        propertyId: kga_tracking_id
      })
      .execute();
    
    
      /**
       * Update the activeUsers component, the Chartjs charts, and the dashboard
       * title whenever the user changes the view.
       */
      viewSelector.on('viewChange', function(data) {
        var title = document.getElementById('view-name');
        title.innerHTML = data.property.name + ' (' + data.view.name + ')';
    
        // Start tracking active users for this view.
        activeUsers.set(data).execute();
    
        // Render all the of charts for this view.
        renderWeekOverWeekChart(data.ids);
        renderYearOverYearChart(data.ids);
        renderTopBrowsersChart(data.ids);
        renderTopCountriesChart(data.ids);
      });
    
    
      /**
       * Draw the a chart.js line chart with data from the specified view that
       * overlays session data for the current week over session data for the
       * previous week.
       */
      function renderWeekOverWeekChart(ids) {
    
        // Adjust `now` to experiment with different days, for testing only...
        var now = moment(); // .subtract(3, 'day');
    
        var thisWeek = query({
          'ids': ids,
          'dimensions': 'ga:date,ga:nthDay',
          'metrics': 'ga:sessions',
          'start-date': moment(now).subtract(1, 'day').day(0).format('YYYY-MM-DD'),
          'end-date': moment(now).format('YYYY-MM-DD')
        });
    
        var lastWeek = query({
          'ids': ids,
          'dimensions': 'ga:date,ga:nthDay',
          'metrics': 'ga:sessions',
          'start-date': moment(now).subtract(1, 'day').day(0).subtract(1, 'week')
              .format('YYYY-MM-DD'),
          'end-date': moment(now).subtract(1, 'day').day(6).subtract(1, 'week')
              .format('YYYY-MM-DD')
        });
    
        Promise.all([thisWeek, lastWeek]).then(function(results) {
    
          var data1 = results[0].rows.map(function(row) { return +row[2]; });
          var data2 = results[1].rows.map(function(row) { return +row[2]; });
          var labels = results[1].rows.map(function(row) { return +row[0]; });
    
          labels = labels.map(function(label) {
            return moment(label, 'YYYYMMDD').format('ddd');
          });
    
          var data = {
            labels : labels,
            datasets : [
              {
                label: 'Last Week',
                fillColor : 'rgba(220,220,220,0.5)',
                strokeColor : 'rgba(220,220,220,1)',
                pointColor : 'rgba(220,220,220,1)',
                pointStrokeColor : '#fff',
                data : data2
              },
              {
                label: 'This Week',
                fillColor : 'rgba(151,187,205,0.5)',
                strokeColor : 'rgba(151,187,205,1)',
                pointColor : 'rgba(151,187,205,1)',
                pointStrokeColor : '#fff',
                data : data1
              }
            ]
          };
    
          new Chart(makeCanvas('chart-1-container')).Line(data);
          generateLegend('legend-1-container', data.datasets);
        });
      }
    
    
      /**
       * Draw the a chart.js bar chart with data from the specified view that
       * overlays session data for the current year over session data for the
       * previous year, grouped by month.
       */
      function renderYearOverYearChart(ids) {
    
        // Adjust `now` to experiment with different days, for testing only...
        var now = moment(); // .subtract(3, 'day');
    
        var thisYear = query({
          'ids': ids,
          'dimensions': 'ga:month,ga:nthMonth',
          'metrics': 'ga:users',
          'start-date': moment(now).date(1).month(0).format('YYYY-MM-DD'),
          'end-date': moment(now).format('YYYY-MM-DD')
        });
    
        var lastYear = query({
          'ids': ids,
          'dimensions': 'ga:month,ga:nthMonth',
          'metrics': 'ga:users',
          'start-date': moment(now).subtract(1, 'year').date(1).month(0)
              .format('YYYY-MM-DD'),
          'end-date': moment(now).date(1).month(0).subtract(1, 'day')
              .format('YYYY-MM-DD')
        });
    
        Promise.all([thisYear, lastYear]).then(function(results) {
          var data1 = results[0].rows.map(function(row) { return +row[2]; });
          var data2 = results[1].rows.map(function(row) { return +row[2]; });
          var labels = ['Jan','Feb','Mar','Apr','May','Jun',
                        'Jul','Aug','Sep','Oct','Nov','Dec'];
    
          // Ensure the data arrays are at least as long as the labels array.
          // Chart.js bar charts don't (yet) accept sparse datasets.
          for (var i = 0, len = labels.length; i < len; i++) {
            if (data1[i] === undefined) data1[i] = null;
            if (data2[i] === undefined) data2[i] = null;
          }
    
          var data = {
            labels : labels,
            datasets : [
              {
                label: 'Last Year',
                fillColor : 'rgba(220,220,220,0.5)',
                strokeColor : 'rgba(220,220,220,1)',
                data : data2
              },
              {
                label: 'This Year',
                fillColor : 'rgba(151,187,205,0.5)',
                strokeColor : 'rgba(151,187,205,1)',
                data : data1
              }
            ]
          };
    
          new Chart(makeCanvas('chart-2-container')).Bar(data);
          generateLegend('legend-2-container', data.datasets);
        })
        .catch(function(err) {
          console.error(err.stack);
        });
      }
    
    
      /**
       * Draw the a chart.js doughnut chart with data from the specified view that
       * show the top 5 browsers over the past seven days.
       */
      function renderTopBrowsersChart(ids) {
    
        query({
          'ids': ids,
          'dimensions': 'ga:browser',
          'metrics': 'ga:pageviews',
          'sort': '-ga:pageviews',
          'max-results': 5
        })
        .then(function(response) {
    
          var data = [];
          var colors = ['#4D5360','#949FB1','#D4CCC5','#E2EAE9','#F7464A'];
    
          response.rows.forEach(function(row, i) {
            data.push({ value: +row[1], color: colors[i], label: row[0] });
          });
    
          new Chart(makeCanvas('chart-3-container')).Doughnut(data);
          generateLegend('legend-3-container', data);
        });
      }
    
    
      /**
       * Draw the a chart.js doughnut chart with data from the specified view that
       * compares sessions from mobile, desktop, and tablet over the past seven
       * days.
       */
      function renderTopCountriesChart(ids) {
        query({
          'ids': ids,
          'dimensions': 'ga:country',
          'metrics': 'ga:sessions',
          'sort': '-ga:sessions',
          'max-results': 5
        })
        .then(function(response) {
    
          var data = [];
          var colors = ['#4D5360','#949FB1','#D4CCC5','#E2EAE9','#F7464A'];
    
          response.rows.forEach(function(row, i) {
            data.push({
              label: row[0],
              value: +row[1],
              color: colors[i]
            });
          });
    
          new Chart(makeCanvas('chart-4-container')).Doughnut(data);
          generateLegend('legend-4-container', data);
        });
      }
    
    
      /**
       * Extend the Embed APIs `gapi.analytics.report.Data` component to
       * return a promise the is fulfilled with the value returned by the API.
       * @param {Object} params The request parameters.
       * @return {Promise} A promise.
       */
      function query(params) {
        return new Promise(function(resolve, reject) {
          var data = new gapi.analytics.report.Data({query: params});
          data.once('success', function(response) { resolve(response); })
              .once('error', function(response) { reject(response); })
              .execute();
        });
      }
    
    
      /**
       * Create a new canvas inside the specified element. Set it to be the width
       * and height of its container.
       * @param {string} id The id attribute of the element to host the canvas.
       * @return {RenderingContext} The 2D canvas context.
       */
      function makeCanvas(id) {
        var container = document.getElementById(id);
        var canvas = document.createElement('canvas');
        var ctx = canvas.getContext('2d');
    
        container.innerHTML = '';
        canvas.width = container.offsetWidth;
        canvas.height = container.offsetHeight;
        container.appendChild(canvas);
    
        return ctx;
      }
    
    
      /**
       * Create a visual legend inside the specified element based off of a
       * Chart.js dataset.
       * @param {string} id The id attribute of the element to host the legend.
       * @param {Array.<Object>} items A list of labels and colors for the legend.
       */
      function generateLegend(id, items) {
        var legend = document.getElementById(id);
        legend.innerHTML = items.map(function(item) {
          var color = item.color || item.fillColor;
          var label = item.label;
          return '<li><i style="background:' + color + '"></i>' + label + '</li>';
        }).join('');
      }
    
    
      // Set some global Chart.js defaults.
      Chart.defaults.global.animationSteps = 60;
      Chart.defaults.global.animationEasing = 'easeInOutQuart';
      Chart.defaults.global.responsive = true;
      Chart.defaults.global.maintainAspectRatio = false;
    
    });
};

kotti_google_analytics/__init__.py
# -*- coding: utf-8 -*-

"""
Created on 2016-06-18
:author: Oshane Bailey (b4.oshany@gmail.com)
"""

from kotti.resources import File
from kotti.views.slots import assign_slot
from pyramid.i18n import TranslationStringFactory
from kotti_controlpanel.util import get_setting


controlpanel_id = 'kotti_google_analytics'

_ = TranslationStringFactory(controlpanel_id)


class AnayticsDefault(object):
    tracking_id = None
    client_secret = None
    client_id = None
    access_token = None
    refresh_token = None
    identity = None


def kotti_configure(settings):
    """ Add a line like this to you .ini file::

            kotti.configurators =
                kotti_google_analytics.kotti_configure

        to enable the ``kotti_google_analytics`` add-on.

    :param settings: Kotti configuration dictionary.
    :type settings: dict
    """

    settings['pyramid.includes'] += ' kotti_google_analytics'
    settings['kotti.populators'] += ' kotti_google_analytics.populate.populate'
    settings['kotti.alembic_dirs'] += ' kotti_google_analytics:alembic'
    settings['kotti.fanstatic.view_needed'] += ' kotti_google_analytics.fanstatic.css_and_js'
    assign_slot('analytics-code', 'belowcontent')


def includeme(config):
    """ Don't add this to your ``pyramid_includes``, but add the
    ``kotti_configure`` above to your ``kotti.configurators`` instead.

    :param config: Pyramid configurator object.
    :type config: :class:`pyramid.config.Configurator`
    """

    config.add_translation_dirs('kotti_google_analytics:locale')
    AnayticsDefault.tracking_id = \
        config.registry.settings.get("kotti_google_analytics.tracking_id", None)
    config.add_static_view('static-kotti_google_analytics', 'kotti_google_analytics:static')

    config.scan(__name__)
